cmd/govim/internal/fswatcher/os_others.go

```diff
@@ -13,49 +16,157 @@ type fswatcher struct {
 	eventCh chan Event
 	errCh   chan error
 	mw      *fsnotify.Watcher
+	logf    logFn
 }
 
-func New(gomodpath string, tomb *tomb.Tomb) (*FSWatcher, error) {
+func New(root string, filter watchFilterFn, logf logFn, tomb *tomb.Tomb) (*FSWatcher, error) {
 	mw, err := fsnotify.NewWatcher()
 	if err != nil {
 		return nil, fmt.Errorf("failed to create new watcher: %v", err)
 	}
 
+	// activeWatchers is a map keyed by directory paths for directories that has been
+	// added to watch. The bool value is used to dedup multiple remove events for the same
+	// directory from fsnotify (inotify sends two events when removing a dir, "DELETE_SELF"
+	// and "DELETE,ISDIR" where the difference isn't exposed by fsnotify).
+	activeWatches := make(map[string]bool)
 	eventCh := make(chan Event)
+
+	w := &FSWatcher{&fswatcher{
+		eventCh: eventCh,
+		errCh:   mw.Errors,
+		mw:      mw,
+		logf:    logf,
+	}}
+
+	var populateWatchers func(string) ([]string, error)
+	populateWatchers = func(initPath string) (files []string, err error) {
+		return files, filepath.Walk(initPath, func(path string, info os.FileInfo, err error) error {
```

Comment:
I'm not such a fan of `files` being populated as a side effect of the second return value here; it hurts readability to my mind. 

Can `populateWatchers` be a method on `fswatcher` - would that tidy things up a bit?

Reply:
Fixed as per our Slack conversation & moved `populatedWatchers` into `fswatcher`.